### PR TITLE
logging: include information about contextual logging

### PR DESCRIPTION
**Pull request by pohly**
In Kubernetes 1.24, the necessary infrastructure got added. Now developers can
take advantage of it.

### CODE REVIEW

File: contributors/devel/sig-instrumentation/migration-to-structured-logging.md [hunk 1]
```diff
@@ -168,47 +172,128 @@ type KMetadata interface {
...
+that developers and reviewers need to be aware of.
+
 ## Migration
 
-1. Change log functions to structured equivalent
+1. Find code locations that need to be changed:
+
+   - `(cd hack/tools && go install k8s.io/klog/hack/tools/logcheck)`
+   - structured logging: `$GOPATH/bin/logcheck -check-structured ./pkg/controller/...`
+   - contextual logging: `$GOPATH/bin/logcheck -check-contextual ./pkg/scheduler/...`
```
Comment (pohly) [inline contributors/devel/sig-instrumentation/migration-to-structured-logging.md:240]: This depends on https://github.com/kubernetes/kubernetes/pull/108159

**Comment by pohly** [comment]
/hold

Code PRs needs to be merged first.

**Comment by pohly** [comment]
/hold cancel
/assign @serathius 

Code PRs are all merged, contextual logging is alpha in 1.24. Let's finish the documentation... :grin:

File: contributors/devel/sig-instrumentation/migration-to-structured-logging.md [hunk 2]
```diff
@@ -168,47 +172,128 @@ type KMetadata interface {
...
+The KEP has further instructions about the [transition to contextual
+logging](https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/3077-contextual-logging#transition). It
+also lists several
+[pitfalls](https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/3077-contextual-logging#pitfalls-during-usage)
+that developers and reviewers need to be aware of.
+
 ## Migration
 
-1. Change log functions to structured equivalent
+1. Find code locations that need to be changed:
```
Comment (serathius) [inline contributors/devel/sig-instrumentation/migration-to-structured-logging.md:236]: I think that this tool would be better to verify the results than find code locations.

File: contributors/devel/sig-instrumentation/migration-to-structured-logging.md [hunk 3]
```diff
@@ -168,47 +172,128 @@ type KMetadata interface {
...
+### Using ErrorS or logger.Error
 
 With `klog` structured logging borrowing the interface from [logr] it also inherits it's differences in semantic of
 error function. Logs generated by `ErrorS` command may be enhanced with additional debug information
 (such as stack traces) or be additionally sent to special error recording tools. Errors should be used to indicate
 unexpected behaviours in code, like unexpected errors returned by subroutine function calls.
+In contrast to info log calls, error log calls always record the log entry, regardless of the current verbosity
+settings.
 
-Calling `ErrorS` with `nil` as error is semi-acceptable if there is error condition that deserves a stack trace at this
```
[the remark this answers is not in the document]
Comment (serathius) [reply contributors/devel/sig-instrumentation/migration-to-structured-logging.md:204]: When writing this document we wanted a clear way for users to decide whether they should use Info or Error. An error that doesn't happen during routine operation and deserves a stacktrace was a distinction proposed by @thockin. In the end didn't implement writing stacktrace.

**Comment by thockin** [comment]
I think there will always be times when I have an "error situation"
(something bad happened, but I am not reporting it further up-stack - maybe
I will retry or something) but I do not have a literal `error` type
variable.  `log.ErrorS()` is still appropriate, IMO

File: contributors/devel/sig-instrumentation/migration-to-structured-logging.md (hunk 2 above)
Comment (pohly) [reply contributors/devel/sig-instrumentation/migration-to-structured-logging.md:236]: > I think that this tool would be better to verify the results than find code locations.

Then how do you suggest that developers should find code locations instead? Using "git grep" easily leads to false positives. The logcheck tool is more precise.

> Could make it part of Verify log output?

That section is about something else (the output that gets generated, not how it gets generated). I can add a "Preventing regressions" section and there describe how we use `hack/logcheck.conf` to ensure that undesirable calls do not get added back into already converted packages.

**Comment by pohly** [comment]
> I think there will always be times when I have an "error situation"
> (something bad happened, but I am not reporting it further up-stack - maybe
> I will retry or something) but I do not have a literal `error` type
> variable.  `log.ErrorS()` is still appropriate, IMO

And that's why both `klog.ErrorS` and `logger.Error` allow the `err` parameter to be nil, so it's not just "semi-acceptable" (old wording, now removed) but encouraged to use the error functions for error situations.

File: contributors/devel/sig-instrumentation/migration-to-structured-logging.md [hunk 4]
```diff
@@ -168,47 +172,129 @@ type KMetadata interface {
...
+- [`klog.Background`](https://pkg.go.dev/k8s.io/klog/v2#Background)
+- [`klog.TODO`](https://pkg.go.dev/k8s.io/klog/v2#TODO)
+
+The preferred approach is to retrieve the instance with `klog.FromContext` from
+a `ctx context` parameter. If a function or method does not have one, consider
+adding it. This API change then implies that all callers also need to be
+updated. If there are any `context.TODO` calls in the modified functions,
+replace with the new `ctx` parameter.
+
+In performance critical code it may be faster to add a `logger klog.Logger`
```
Comment (logicalhan) [inline contributors/devel/sig-instrumentation/migration-to-structured-logging.md:201]: Do you have any hard numbers on this?

Comment (pohly) [reply contributors/devel/sig-instrumentation/migration-to-structured-logging.md:201]: I did some microbenchmarking for the KEP: https://github.com/kubernetes/enhancements/pull/3078#discussion_r784199814

Those tests are pending for Kubernetes: https://github.com/kubernetes/kubernetes/pull/109194/commits

**Review by chrisnegus — COMMENTED** [review]
/lgtm

**Comment by nate-double-u** [comment]
Not sure if I'm able to approve in this repo, but this looks good to me.
/approve

**Comment by pohly** [comment]
/assign @serathius 

For final review and approval.

**Pull request merged**